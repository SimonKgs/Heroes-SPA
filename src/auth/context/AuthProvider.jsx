import { useReducer } from "react"
import { AuthContext } from "./AuthContext"
import { authReducer } from "./authReducer"

import { types } from "../types/types"

//  ahora inicio en el init lo dejo de ejemplo, lo modifico por un {}
// const initialState = {
//     logged: false,
// }

const init = () => {
    const user = JSON.parse( localStorage.getItem('user') );

    return {
        logged: !!user,
        user: user
    }
}

export const AuthProvider = ({ children }) => {

    const [ authState, dispatchAuth ] = useReducer(authReducer, {}, init);
    
    const onLogin = ( name = '' ) => {
        
        const user = {
            id: '1',
            name
        }

        const action = {
            type: types.login,
            payload: user,
        }

        localStorage.setItem('user', JSON.stringify(user));
        
        dispatchAuth(action)
    }

    const onLogout = () => {
        //  le pasamos un objeto vacio en lugar de una cadena vacia para evitar que el parse del init rompa la aplicacion si se recarga la pagina
        //  localStorage.setItem('user', JSON.stringify({ user: ''}));
        //  dejo de ejemplo pero mas facil y sin errores removemos el item, al encontrarse el logged !!user recordemos que pasa el null a true y lo devuelvo a false
        //  con ello no intenta hacer el parse del user pues lo borro
        localStorage.removeItem('user');
        dispatchAuth( { type: types.logout } );
    }

    return (
        <AuthContext.Provider value={{ ...authState, onLogin, onLogout }}>
            {children}
        </AuthContext.Provider>

    )
}
