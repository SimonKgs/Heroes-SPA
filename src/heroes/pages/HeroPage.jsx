import React, { useMemo } from "react";
import { Navigate, useNavigate, useParams } from "react-router-dom";
import { getHeroById } from "../helpers";

export const HeroPage = () => {
  const { id, ...rest } = useParams();
  const navigate = useNavigate();

  //  useMemo para mejorar el rendimiento en caso de que se fuese a renderizar de nuevo el componente por algun cambio
  //  en este caso no lo hara, pero asi me acostumbro a tener memorizadas las funciones internas de los componentes
  const hero = useMemo(() => getHeroById(id), [id]);

  //  A diferencia del Navigate que es un componente, el useNavigate es un hook, como una funcion y permite pasarle el -1
  //  !leer sobre ello en la doc de react-router 6
  const onNavigateBack = () => {
    hero.publisher === "DC Comics" ? navigate("/dc") : navigate("/marvel");
  };

  if (!hero) {
    return <Navigate to="/marvel" />;
    // return <>404 - not found</>
  }

  return (
    <div className="row mt-5">
      <div className="col-4">
        <img
          src={` /assets/heroes/${id}.jpg`}
          alt={hero.superhero}
          className="img-thumbnail animate__animated animate__flash animate__slower"
        />
      </div>
      <div className="col-8">
        <h3>{hero.superhero}</h3>
        <ul className="list-group list-group-flush">
          <li className="list-group-item">
            <b>Alter ego:</b> {hero.alter_ego}
          </li>
          <li className="list-group-item">
            <b>Publisher:</b> {hero.publisher}
          </li>
          <li className="list-group-item">
            <b>First appearence:</b> {hero.first_appearance}
          </li>
        </ul>

        <h5 className="mt-3">Characters</h5>
        <p>{hero.characters}</p>

        <button className="btn btn-outline-primary" onClick={onNavigateBack}>
          back
        </button>
      </div>
    </div>
  );
};
