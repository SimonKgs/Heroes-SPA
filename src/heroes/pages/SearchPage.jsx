import { useLocation, useNavigate } from "react-router-dom";

import queryString from "query-string";

import { useForm } from "../../hooks/useForm";
import { HeroCard } from "../components";
import { getHeroByName } from "../helpers";

export const SearchPage = () => {
  const navigate = useNavigate();

  //usamos el location que viene tambien con el router, es un objeto obtenemos de el search y pathname
  const location = useLocation();

  //  paquete para ayudar a desmenbrar la busqueda en la url, todo lo que tenga sera un string
  const { q = "" } = queryString.parse(location.search);

  const heroes = getHeroByName(q);

  const showSearch = q.length === 0;
  const showError = q.length > 0 && heroes.length === 0;

  const { searchText, onInputChange, onCleanForm } = useForm({
    searchText: q,
  });

  const onSearchSubmit = (e) => {
    e.preventDefault();

    //  navigate(``), apunta a la ruta en la que me encuentro
    navigate(`?q=${searchText.toLowerCase()}`);
  };

  return (
    <>
      <h1>Search</h1>
      <hr />
      <div className="row">
        <div className="col-5">
          <h4>Searching</h4>
          <hr />
          <form onSubmit={onSearchSubmit}>
            <input
              aria-label="input"
              type="text"
              placeholder="search a hero"
              className="form-control"
              name="searchText"
              autoComplete="off"
              value={searchText}
              onChange={onInputChange}
            ></input>
            <button className="btn btn-primary mt-2">Search</button>
          </form>
        </div>
        <div className="col-7">
          <h4>Results</h4>
          <hr />
          <div
            className="alert alert-primary animate__animated animate__fadeIn"
            style={{ display: showSearch ? "" : "none" }}
            aria-label="searchDiv"
          >
            Search a hero
          </div>
          <div
            className="alert alert-danger animate__animated animate__fadeIn"
            style={{ display: showError ? "" : "none" }}
            aria-label="searchError"
          >
            Not found <b>{q}</b>
          </div>
          {/* {
                Lo de arriba es lo mismo pero queda mas legible
              ( showSearch )
              ? <div className="alert alert-primary">Search a hero</div>
              : (heroes.length === 0) && <div className="alert alert-danger">Not found <b>{ q }</b></div>
            } */}
          {heroes.map((hero) => (
            <div className="mt-3" key={hero.id}>
              <HeroCard {...hero} />
            </div>
          ))}
        </div>
      </div>
    </>
  );
};
