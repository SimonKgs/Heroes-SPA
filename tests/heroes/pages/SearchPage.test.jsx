import { render, screen, fireEvent } from '@testing-library/react'
import { MemoryRouter, useLocation } from 'react-router-dom'
import { AuthContext } from '../../../src/auth/context/AuthContext'
import { SearchPage } from '../../../src/heroes/pages/SearchPage'


const mockedUseNavigate = jest.fn()

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedUseNavigate,
}))


describe('Pruebas en <SearchScreen />', () => {

    const authContext = {
        logged: true,
        user: {
            id: 1,
            name: 'simon'
        }
    }

    beforeEach( () => jest.clearAllMocks() );
    test('Debe de hacer match con el snapshot', () => {


        const { container } = render(
            <MemoryRouter>
                <AuthContext.Provider value={ authContext }>
                    <SearchPage />
                </AuthContext.Provider>
            </MemoryRouter>
        )

        expect( container ).toMatchSnapshot();

    })

    test('Debe de mostrar a batman y el input con el valor del queryString', () => { 
        render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <AuthContext.Provider value={ authContext }>
                    <SearchPage />
                </AuthContext.Provider>
            </MemoryRouter>
        )

        // para los input usamos textbox
        // const input = screen.getByRole('textbox'); 
        //  utilice un aria-label en su lugar para acceder al input que quiero
        const inputValue = screen.getByLabelText('input');
        expect( inputValue.value ).toBe('batman');

        const searchDiv = screen.getByLabelText("searchDiv");
        expect( searchDiv.style.display ).toBe( "none" )
        
        const img = screen.getByRole('img');
        expect( img.src ).toContain("/assets/heroes/dc-batman.jpg")
        
    })

    test('Debe de llamar al navigate con los valores de la busqueda', () => { 
        render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <AuthContext.Provider value={ authContext }>
                    <SearchPage />
                </AuthContext.Provider>
            </MemoryRouter>
        )
        
        //  obtnemos el input y lo lanzamos con change dandole el nombre y el valor que tomara la caja
        //  no es necesario como yo lo hice pues mi texbox toma el valor de la url en las entries 
        //  pero con ello probamos tambien el propio input
        const input = screen.getByRole('textbox');
        fireEvent.change( input, { target: { name: 'searchText', value: 'batman'}})
        
        //  tambien podria haber simulado el envio del formulario pero lo hice con el boton, el funcionamiento es el mismo
        //  const form = screen.getByRole('form') !! el form no lo coge jest, ponerle un aria label 
        //  fireEvent.submit( form );
        const searchButton = screen.getByRole('button');
        fireEvent.click(searchButton)
        expect( mockedUseNavigate ).toHaveBeenCalledWith("?q=batman")
        
        expect( screen.getByText('Batman').innerHTML ).toBe('Batman')
     });

     test('Debe de mostrar el error en caso de que no haya encontrado un heroe', () => { 
        render(
            <MemoryRouter initialEntries={['/search?q=asa']}>
                <AuthContext.Provider value={ authContext }>
                    <SearchPage />
                </AuthContext.Provider>
            </MemoryRouter>
        )

        const searchError = screen.getByLabelText("searchError");
        expect( searchError.style.display ).toBe( "" )


      })
})