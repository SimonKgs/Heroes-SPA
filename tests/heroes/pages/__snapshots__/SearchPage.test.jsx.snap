// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen /> Debe de hacer match con el snapshot 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form>
        <input
          aria-label="input"
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="search a hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-primary mt-2"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        aria-label="searchDiv"
        class="alert alert-primary animate__animated animate__fadeIn"
      >
        Search a hero
      </div>
      <div
        aria-label="searchError"
        class="alert alert-danger animate__animated animate__fadeIn"
        style="display: none;"
      >
        Not found 
        <b />
      </div>
    </div>
  </div>
</div>
`;
